File: backend/app/models/character.py
from sqlalchemy import Column, Integer, String, DateTime, Text, ForeignKey, Enum
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import JSONB
from app.core.database import Base
import enum


class CharacterRole(str, enum.Enum):
    PROTAGONIST = "protagonist"  # 主角
    ANTAGONIST = "antagonist"  # 反派
    SUPPORTING = "supporting"  # 配角
    MINOR = "minor"  # 次要角色


class Character(Base):
    __tablename__ = "characters"

    id = Column(Integer, primary_key=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=False)

    # 基础信息
    name = Column(String, nullable=False)
    age = Column(Integer, nullable=True)
    gender = Column(String, nullable=True)
    appearance = Column(Text, nullable=True)  # 外貌描述
    identity = Column(String, nullable=True)  # 身份
    hometown = Column(String, nullable=True)  # 籍贯

    # 核心人设
    role = Column(Enum(CharacterRole), default=CharacterRole.SUPPORTING)
    personality = Column(Text, nullable=True)  # 性格标签，JSON格式
    core_motivation = Column(Text, nullable=True)  # 核心动机
    fears = Column(Text, nullable=True)  # 恐惧
    desires = Column(Text, nullable=True)  # 欲望

    # 人物弧光（保留旧字段用于兼容）
    initial_state = Column(Text, nullable=True)  # 起始状态（已弃用）
    turning_point = Column(Text, nullable=True)  # 转折点（已弃用）
    final_state = Column(Text, nullable=True)  # 最终状态（已弃用）

    # 人物弧光 - 多条转变记录（新字段）
    character_arcs = Column(JSONB, nullable=True)  # [{"period": "", "event": "", "before": "", "after": ""}]

    # 语音风格（保留旧字段用于兼容）
    speech_style = Column(Text, nullable=True)  # 语音风格描述（已弃用）
    sample_dialogue = Column(Text, nullable=True)  # 样本对话（已弃用）

    # 语音风格 - 多场景风格记录（新字段）
    voice_styles = Column(JSONB, nullable=True)  # [{"target": "", "scenario": "", "style": "", "sample": ""}]

    # 头像
    avatar = Column(String, nullable=True)

    # 统计
    appearance_count = Column(Integer, default=0)  # 出场次数

    # 时间戳
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # 关系
    project = relationship("Project", back_populates="characters")

File: backend-java/src/main/java/com/novel/ai/model/entity/Character.java
package com.novel.ai.model.entity;

import java.time.LocalDateTime;

/**
 * 人物实体类
 * 用于存储小说项目中的人物信息，包括基础信息、核心人设、人物弧光、语音风格等
 */
public class Character {

    // 基础字段
    private Integer id;
    private Integer projectId;

    // 基础信息
    private String name;
    private Integer age;
    private String gender;
    private String appearance;      // 外貌描述
    private String identity;        // 身份
    private String hometown;        // 籍贯

    // 核心人设
    private String role;            // 角色类型：protagonist/antagonist/supporting/minor
    private String personality;    // 性格标签，JSON格式
    private String coreMotivation;  // 核心动机
    private String fears;           // 恐惧
    private String desires;         // 欲望

    // 人物弧光 - JSON格式存储多条转变记录
    // 示例：[{"period": "初期", "event": "关键事件", "before": "转变前", "after": "转变后"}]
    private String characterArcs;

    // 语音风格 - JSON格式存储多场景风格记录
    // 示例：[{"target": "对话对象", "scenario": "场景", "style": "风格", "sample": "样本对话"}]
    private String voiceStyles;

    // 时间戳
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    // 构造函数
    public Character() {}

    public Character(Integer projectId, String name) {
        this.projectId = projectId;
        this.name = name;
        this.role = "supporting";  // 默认为配角
    }

    // Getter和Setter方法
    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public Integer getProjectId() {
        return projectId;
    }

    public void setProjectId(Integer projectId) {
        this.projectId = projectId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Integer getAge() {
        return age;
    }

    public void setAge(Integer age) {
        this.age = age;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getAppearance() {
        return appearance;
    }

    public void setAppearance(String appearance) {
        this.appearance = appearance;
    }

    public String getIdentity() {
        return identity;
    }

    public void setIdentity(String identity) {
        this.identity = identity;
    }

    public String getHometown() {
        return hometown;
    }

    public void setHometown(String hometown) {
        this.hometown = hometown;
    }

    public String getRole() {
        return role;
    }

    public void setRole(String role) {
        this.role = role;
    }

    public String getPersonality() {
        return personality;
    }

    public void setPersonality(String personality) {
        this.personality = personality;
    }

    public String getCoreMotivation() {
        return coreMotivation;
    }

    public void setCoreMotivation(String coreMotivation) {
        this.coreMotivation = coreMotivation;
    }

    public String getFears() {
        return fears;
    }

    public void setFears(String fears) {
        this.fears = fears;
    }

    public String getDesires() {
        return desires;
    }

    public void setDesires(String desires) {
        this.desires = desires;
    }

    public String getCharacterArcs() {
        return characterArcs;
    }

    public void setCharacterArcs(String characterArcs) {
        this.characterArcs = characterArcs;
    }

    public String getVoiceStyles() {
        return voiceStyles;
    }

    public void setVoiceStyles(String voiceStyles) {
        this.voiceStyles = voiceStyles;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}